Issue parsing MacOS CoreGraphics

```
% cat c.h
#include <CoreGraphics/CoreGraphics.h>
```

```
arocc c.h \
  -nostdlibinc \
  -O0 \
  -idirafter "/nix/store/elided-zig-0.16.0/lib/zig/include" \
  -isystem "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include" \
  -isystem "/opt/homebrew/include" \
  -iframework "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks"
```

Output below.

I think this is possibly related to what @dotcarmen is working on in #971 but I'm not too sure. Debug sessions would indicate that this is failing on actually printing the diagnostic on sections that are very close to the ones that are actually printing in relation to the lack of ability to null the block types, so I'm not too sure if it's something that would be better covered once this work lands, or if all of this is a red herring. 🙂 

## Output

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGPDFArray.h:113:40: error: nullability specifier cannot be applied to non-pointer type 'CGPDFArrayApplierBlock' (aka '_Bool (^)(unsigned long, struct CGPDFObject *, void *)')
    CF_NOESCAPE CGPDFArrayApplierBlock cg_nullable block, void * __nullable info)
                                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGBase.h:352:24: note: expanded from here
#   define cg_nullable __nullable
                       ^
<builtin>:79:20: note: expanded from here
#define __nullable _Nullable
                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGPDFDictionary.h:126:45: error: nullability specifier cannot be applied to non-pointer type 'CGPDFDictionaryApplierBlock' (aka '_Bool (^)(const char *, struct CGPDFObject *, void *)')
    CF_NOESCAPE CGPDFDictionaryApplierBlock cg_nullable block, void * __nullable info)
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGBase.h:352:24: note: expanded from here
#   define cg_nullable __nullable
                       ^
<builtin>:79:20: note: expanded from here
#define __nullable _Nullable
                   ^
panic: reached unreachable code
/Users/elided/arocc/src/aro/TypeStore.zig:194:36: 0x1010eabfb in type (arocc)
            .declarator_combine => unreachable,
                                   ^
/Users/elided/arocc/src/aro/TypeStore.zig:432:38: 0x101129dbf in base (arocc)
        while (true) switch (cur.type(comp)) {
                                     ^
/Users/elided/arocc/src/aro/TypeStore.zig:454:34: 0x101170803 in get__anon_61198 (arocc)
        const base_type = qt.base(comp).type;
                                 ^
/Users/elided/arocc/src/aro/TypeStore.zig:460:22: 0x10116fefb in is__anon_61297 (arocc)
        return qt.get(comp, tag) != null;
                     ^
/Users/elided/arocc/src/aro/TypeStore.zig:1327:35: 0x10116b7b7 in printPrologue (arocc)
                if (!block.func.is(comp, .func)) unreachable;
                                  ^
/Users/elided/arocc/src/aro/TypeStore.zig:1280:33: 0x101169dff in print (arocc)
        _ = try qt.printPrologue(comp, false, w);
                                ^
/Users/elided/arocc/src/aro/Parser.zig:508:17: 0x101180993 in formatQualType (arocc)
    try qt.print(p.comp, w);
                ^
/Users/elided/arocc/src/aro/Parser.zig:477:45: 0x10118109b in formatArgs__anon_62907 (arocc)
            QualType => try p.formatQualType(w, fmt[i..], arg),
                                            ^
/Users/elided/arocc/src/aro/Parser.zig:451:17: 0x1011814c3 in err__anon_62564 (arocc)
    p.formatArgs(&allocating.writer, diagnostic.fmt, args) catch return error.OutOfMemory;
                ^
/Users/elided/arocc/src/aro/TypeStore.zig:3027:33: 0x101308af3 in finishQuals (arocc)
                try b.parser.err(tok, .invalid_nullability, .{qt});
                                ^
/Users/elided/arocc/src/aro/Parser.zig:3727:39: 0x10130934f in declarator (arocc)
        d.qt = try builder.finishQuals(block_qt);
                                      ^
/Users/elided/arocc/src/aro/Parser.zig:3757:36: 0x101309b2f in declarator (arocc)
        var res = (try p.declarator(special_marker, kind)) orelse {
                                   ^
/Users/elided/arocc/src/aro/Parser.zig:4077:29: 0x10130f5f7 in paramDecls (arocc)
        if (try p.declarator(param_qt, .param)) |some| {
                            ^
/Users/elided/arocc/src/aro/Parser.zig:3981:29: 0x10130d647 in directDeclarator (arocc)
        if (try p.paramDecls()) |params| {
                            ^
/Users/elided/arocc/src/aro/Parser.zig:3746:38: 0x10130987b in declarator (arocc)
        d.qt = try p.directDeclarator(&d, kind);
                                     ^
/Users/elided/arocc/src/aro/Parser.zig:2117:31: 0x1013146a3 in initDeclarator (arocc)
        .d = (try p.declarator(decl_spec.qt, .normal)) orelse return null,
                              ^
/Users/elided/arocc/src/aro/Parser.zig:1209:39: 0x101190cf3 in decl (arocc)
    var init_d = (try p.initDeclarator(&decl_spec, attr_buf_top, decl_node)) orelse {
                                      ^
/Users/elided/arocc/src/aro/Parser.zig:1062:16: 0x101189807 in parseOrNextDecl__anon_62456 (arocc)
    return func(p) catch |er| switch (er) {
               ^
/Users/elided/arocc/src/aro/Parser.zig:958:34: 0x10117b48b in parse (arocc)
        if (try p.parseOrNextDecl(decl)) continue;
                                 ^
/Users/elided/arocc/src/aro/Driver.zig:1454:28: 0x10107a71f in processSource__anon_32571 (arocc)
    var tree = try pp.parse();
                           ^
/Users/elided/arocc/src/aro/Driver.zig:1244:28: 0x10107ebe3 in main__anon_30342 (arocc)
        try d.processSource(tc, source, builtin_macros, user_macros, fast_exit, asm_gen_fn);
                           ^
/Users/elided/arocc/src/main.zig:92:16: 0x1010649d3 in main (arocc)
    driver.main(&toolchain, args, fast_exit, assembly_backend.genAsm) catch |er| switch (er) {
               ^
/nix/store/elided-zig-0.16.0/lib/zig/std/start.zig:699:88: 0x101063c17 in callMain (arocc)
    if (fn_info.params[0].type.? == std.process.Init.Minimal) return wrapMain(root.main(.{
                                                                                       ^
???:?:?: 0x18aab3dff in start (/usr/lib/dyld)
```

Some selective debugger frames:

```
    frame #6: 0x0000000100214994 arocc`aro.Parser.formatQualType(p=0x000000016fdf3a40, w=0x000000016fdecef0, fmt=(ptr = "nullability specifier cannot be applied to non-pointer type {qt}", len = 64), qt=(bits = 65320)) at Parser.zig:508:17
    frame #7: 0x000000010021509c arocc`aro.Parser.formatArgs__anon_62907(p=0x000000016fdf3a40, w=0x000000016fdecef0, fmt=(ptr = "nullability specifier cannot be applied to non-pointer type {qt}", len = 64), args=<unavailable>) at Parser.zig:477:45
    frame #8: 0x00000001002154c4 arocc`aro.Parser.err__anon_62564(p=0x000000016fdf3a40, tok_i=110442, diagnostic=aro.Parser.Diagnostic @ 0x00000001006b09d8, args=struct { aro.TypeStore.QualType } @ 0x000000016fded10c) at Parser.zig:451:17
    frame #9: 0x000000010039caf4 arocc`aro.TypeStore.Builder.finishQuals(b=<unavailable>, qt=(bits = 65320)) at TypeStore.zig:3027:33
```

@Vexu new output after squashing #1048 onto master:

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:233:116: error: use of undeclared identifier 'CFString'
CFStringRef CFStringCreateWithFormat(CFAllocatorRef alloc, CFDictionaryRef formatOptions, CFStringRef format, ...) CF_FORMAT_FUNCTION(3,4);
                                                                                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:236:142: error: use of undeclared identifier 'CFString'
CFStringRef CFStringCreateWithFormatAndArguments(CFAllocatorRef alloc, CFDictionaryRef formatOptions, CFStringRef format, va_list arguments) CF_FORMAT_FUNCTION(3,0);
                                                                                                                                             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:239:251: error: use of undeclared identifier 'CFString'
CFStringRef CFStringCreateStringWithValidatedFormat(CFAllocatorRef alloc, CFDictionaryRef formatOptions, CFStringRef validFormatSpecifiers, CFStringRef format, CFErrorRef *errorPtr, ...) API_AVAILABLE(macos(13.0), ios(16.0), watchos(8.0), tvos(8.0)) CF_FORMAT_FUNCTION(3, 6) CF_SWIFT_UNAVAILABLE("Use string interpolations instead");
                                                                                                                                                                                                                                                          ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:242:277: error: use of undeclared identifier 'CFString'
CFStringRef CFStringCreateStringWithValidatedFormatAndArguments(CFAllocatorRef alloc, CFDictionaryRef formatOptions, CFStringRef validFormatSpecifiers, CFStringRef format, va_list arguments, CFErrorRef *errorPtr) API_AVAILABLE(macos(13.0), ios(16.0), watchos(8.0), tvos(8.0)) CF_FORMAT_FUNCTION(3, 0) CF_SWIFT_UNAVAILABLE("Use string interpolations instead");
                                                                                                                                                                                                                                                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:582:113: error: use of undeclared identifier 'CFString'
void CFStringAppendFormat(CFMutableStringRef theString, CFDictionaryRef formatOptions, CFStringRef format, ...) CF_FORMAT_FUNCTION(3,4);
                                                                                                                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:585:139: error: use of undeclared identifier 'CFString'
void CFStringAppendFormatAndArguments(CFMutableStringRef theString, CFDictionaryRef formatOptions, CFStringRef format, va_list arguments) CF_FORMAT_FUNCTION(3,0);
                                                                                                                                          ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFString.h:164:55: note: expanded from here
#define CF_FORMAT_FUNCTION(F,A) __attribute__((format(CFString, F, A)))
                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGPDFArray.h:113:40: error: nullability specifier cannot be applied to non-pointer type 'CGPDFArrayApplierBlock' (aka '_Bool (^)(unsigned long, struct CGPDFObject *, void *)')
    CF_NOESCAPE CGPDFArrayApplierBlock cg_nullable block, void * __nullable info)
                                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGBase.h:352:24: note: expanded from here
#   define cg_nullable __nullable
                       ^
<builtin>:79:20: note: expanded from here
#define __nullable _Nullable
                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGPDFDictionary.h:126:45: error: nullability specifier cannot be applied to non-pointer type 'CGPDFDictionaryApplierBlock' (aka '_Bool (^)(const char *, struct CGPDFObject *, void *)')
    CF_NOESCAPE CGPDFDictionaryApplierBlock cg_nullable block, void * __nullable info)
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/System/Library/Frameworks/CoreGraphics.framework/Headers/CGBase.h:352:24: note: expanded from here
#   define cg_nullable __nullable                                                                                                                                                     ^
<builtin>:79:20: note: expanded from here
#define __nullable _Nullable
                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_printf.h:34:43: error: use of undeclared identifier '__printf__'
int      printf(const char * __restrict, ...) __printflike(1, 2);                                                                                                                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:245:63: error: use of undeclared identifier '__printf__'
int      fprintf(FILE * __restrict, const char * __restrict, ...) __printflike(2, 3);                                                                                                                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:251:62: error: use of undeclared identifier '__scanf__'
int      fscanf(FILE * __restrict, const char * __restrict, ...) __scanflike(2, 3);
                                                             ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:271:42: error: use of undeclared identifier '__scanf__'
int      scanf(const char * __restrict, ...) __scanflike(1, 2);
                                         ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:280:86: error: use of undeclared identifier '__printf__'
int      sprintf(char * __restrict _LIBC_UNSAFE_INDEXABLE, const char * __restrict, ...) __printflike(2, 3) _LIBC_PTRCHECK_REPLACED(snprintf);
                                                                                     ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here            __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:282:68: error: use of undeclared identifier '__scanf__'
int      sscanf(const char * __restrict, const char * __restrict, ...) __scanflike(2, 3);
                                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:292:68: error: use of undeclared identifier '__printf__'
int      vfprintf(FILE * __restrict, const char * __restrict, va_list) __printflike(2, 0);
                                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here            __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:293:48: error: use of undeclared identifier '__printf__'
int      vprintf(const char * __restrict, va_list) __printflike(1, 0);
                                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here            __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:300:91: error: use of undeclared identifier '__printf__'
int      vsprintf(char * __restrict _LIBC_UNSAFE_INDEXABLE, const char * __restrict, va_list) __printflike(2, 0) _LIBC_PTRCHECK_REPLACED(vsnprintf);
                                                                                          ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here            __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:355:47: error: use of undeclared identifier '__scanf__'
int     __svfscanf(FILE *, const char *, va_list) __scanflike(2, 0);
                                              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:439:114: error: use of undeclared identifier '__printf__'
int      snprintf(char * __restrict _LIBC_COUNT(__size) __str, size_t __size, const char * __restrict __format, ...) __printflike(3, 4);
                                                                                                                 ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here            __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:440:85: error: use of undeclared identifier '__scanf__'
int      vfscanf(FILE * __restrict __stream, const char * __restrict __format, va_list) __scanflike(2, 0);
                                                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:441:56: error: use of undeclared identifier '__scanf__'
int      vscanf(const char * __restrict __format, va_list) __scanflike(1, 0);
                                                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here            __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:442:119: error: use of undeclared identifier '__printf__'
int      vsnprintf(char * __restrict _LIBC_COUNT(__size) __str, size_t __size, const char * __restrict __format, va_list) __printflike(3, 0);
                                                                                                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:443:88: error: use of undeclared identifier '__scanf__'
int      vsscanf(const char * __restrict __str, const char * __restrict __format, va_list) __scanflike(2, 0);
                                                                                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:434:29: note: expanded from here
        __attribute__((__format__ (__scanf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:457:48: error: use of undeclared identifier '__printf__'
int     dprintf(int, const char * __restrict, ...) __printflike(2, 3) __OSX_AVAILABLE_STARTING(__MAC_10_7, __IPHONE_4_3);
                                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:458:53: error: use of undeclared identifier '__printf__'
int     vdprintf(int, const char * __restrict, va_list) __printflike(2, 0) __OSX_AVAILABLE_STARTING(__MAC_10_7, __IPHONE_4_3);
                                                    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:475:75: error: use of undeclared identifier '__printf__'
int      asprintf(char *_LIBC_CSTR *__restrict, const char * __restrict, ...) __printflike(2, 3);
                                                                          ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/_stdio.h:482:80: error: use of undeclared identifier '__printf__'
int      vasprintf(char *_LIBC_CSTR *__restrict, const char * __restrict, va_list) __printflike(2, 0);
                                                                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/sys/cdefs.h:430:29: note: expanded from here
        __attribute__((__format__ (__printf__, fmtarg, firstvararg)))
                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/dispatch/object.h:659:27: error: builtin function must be directly called
__attribute__((__format__(printf,2,3)))
                          ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.5.sdk/usr/include/dispatch/object.h:666:27: error: builtin function must be directly called
__attribute__((__format__(printf,2,0)))
                          ^
panic: reached unreachable code
/Users/elided/arocc/src/aro/TypeStore.zig:201:36: 0x100c4dd7b in type (arocc)
            .declarator_combine => unreachable,
                                   ^
/Users/elided/arocc/src/aro/TypeStore.zig:452:38: 0x100c7e43b in base (arocc)
        while (true) switch (cur.type(comp)) {
                                     ^
/Users/elided/arocc/src/aro/TypeStore.zig:479:34: 0x100cc3c0b in get__anon_57163 (arocc)
        const base_type = qt.base(comp).type;
                                 ^
/Users/elided/arocc/src/aro/TypeStore.zig:485:22: 0x100cc3457 in is__anon_57277 (arocc)
        return qt.get(comp, tag) != null;
                     ^
/Users/elided/arocc/src/aro/TypeStore.zig:1317:35: 0x100cbe8df in printPrologue (arocc)
                if (!block.func.is(comp, .func)) unreachable;
                                  ^
/Users/elided/arocc/src/aro/TypeStore.zig:1266:33: 0x100cbd13b in print (arocc)
        _ = try qt.printPrologue(comp, false, w);
                                ^
/Users/elided/arocc/src/aro/Parser.zig:500:17: 0x100cd483f in formatQualType (arocc)
    try qt.print(p.comp, w);
                ^
/Users/elided/arocc/src/aro/Parser.zig:470:45: 0x100cd4f73 in formatArgs__anon_59061 (arocc)
            QualType => try p.formatQualType(w, fmt[i..], arg),
                                            ^
/Users/elided/arocc/src/aro/Parser.zig:444:17: 0x100cd53bf in err__anon_58730 (arocc)
    p.formatArgs(&allocating.writer, diagnostic.fmt, args) catch return error.OutOfMemory;
                ^
/Users/elided/arocc/src/aro/TypeStore.zig:3043:33: 0x100d74997 in finishQuals (arocc)
                try b.parser.err(tok, .invalid_nullability, .{qt});
                                ^
/Users/elided/arocc/src/aro/Parser.zig:3675:39: 0x100dbb797 in declarator (arocc)
        d.qt = try builder.finishQuals(block_qt);
                                      ^
/Users/elided/arocc/src/aro/Parser.zig:3707:36: 0x100dbc0cb in declarator (arocc)
        var res = (try p.declarator(special_marker, kind)) orelse {
                                   ^
/Users/elided/arocc/src/aro/Parser.zig:4028:29: 0x100dd3a07 in paramDecls (arocc)
        if (try p.declarator(param_qt, .param)) |some| {
                            ^
/Users/elided/arocc/src/aro/Parser.zig:3931:29: 0x100dc7fe3 in directDeclarator (arocc)
        if (try p.paramDecls()) |params| {
                            ^
/Users/elided/arocc/src/aro/Parser.zig:3696:38: 0x100dbbe0b in declarator (arocc)
        d.qt = try p.directDeclarator(&d, kind);
                                     ^
/Users/elided/arocc/src/aro/Parser.zig:2155:31: 0x100e368e7 in initDeclarator (arocc)
        .d = (try p.declarator(decl_spec.qt, .normal)) orelse return null,
                              ^
/Users/elided/arocc/src/aro/Parser.zig:1268:39: 0x100e2d3c7 in decl (arocc)
    var init_d = (try p.initDeclarator(&decl_spec, decl_node)) orelse {
                                      ^
/Users/elided/arocc/src/aro/Parser.zig:1122:16: 0x100e42ddf in parseOrNextDecl__anon_58628 (arocc)
    return func(p) catch |er| switch (er) {
               ^
/Users/elided/arocc/src/aro/Parser.zig:1018:34: 0x100ccf2ef in parse (arocc)
        if (try p.parseOrNextDecl(decl)) continue;
                                 ^
/Users/elided/arocc/src/aro/Driver.zig:1456:28: 0x100bcfaf3 in processSource__anon_29828 (arocc)
    var tree = try pp.parse();
                           ^
/Users/elided/arocc/src/aro/Driver.zig:1246:28: 0x100bd401f in main__anon_27624 (arocc)
        try d.processSource(tc, source, builtin_macros, user_macros, fast_exit, asm_gen_fn);
                           ^
/Users/elided/arocc/src/main.zig:92:16: 0x100bb9f33 in main (arocc)
    driver.main(&toolchain, args, fast_exit, assembly_backend.genAsm) catch |er| switch (er) {
               ^
/nix/store/elided-zig-elided/lib/zig/std/start.zig:740:88: 0x100bb915f in callMain (arocc)
    if (fn_info.param_types[0].? == std.process.Init.Minimal) return wrapMain(root.main(.{
                                                                                       ^
???:?:?: 0x18aab3dff in start (/usr/lib/dyld)
zsh: abort      ../../arocc/zig-out/bin/arocc c.h -nostdlibinc -O0 -idirafter  -isystem

```

--- a/src/aro/Attribute/Wip.zig
+++ b/src/aro/Attribute/Wip.zig
@@ -1310,17 +1310,6 @@ fn applyNullability(wip: *Wip) !void {
     const qt = wip.current.qt;
     if (qt.isInvalid()) return;
 
-    const comp = wip.current.parser.comp;
-    var pointer: Type.Pointer = qt.get(comp, .pointer) orelse {
-        if (wip.current.target == null and qt.is(comp, .array)) {
-            attr.used_as_type_attr = false;
-            return;
-        }
-
-        try wip.err(.invalid_nullability, .{qt});
-        return;
-    };
-
     const new_nullability: Type.Pointer.Nullability = switch (attr.name.keyword) {
         .nonnull => .nonnull,
         .null_unspecified => .unspecified,
@@ -1328,13 +1317,32 @@ fn applyNullability(wip: *Wip) !void {
         .nullable => .nullable,
         else => unreachable,
     };
-    if (pointer.nullability != .default and pointer.nullability != new_nullability) {
-        try wip.err(.conflicting_nullability, .{ new_nullability.str(), pointer.nullability.str() });
-        return;
-    }
 
-    pointer.nullability = new_nullability;
-    wip.current.qt = try comp.type_store.put(comp.gpa, .{ .pointer = pointer });
+    const comp = wip.current.parser.comp;
+    var base_type = qt.base(comp).type;
+    switch (base_type) {
+        inline .pointer, .block => |*bt, tag| {
+            if (bt.nullability != .default and bt.nullability != new_nullability) {
+                try wip.err(.conflicting_nullability, .{ new_nullability.str(), bt.nullability.str() });
+                return;
+            }
+
+            bt.nullability = new_nullability;
+            wip.current.qt = try comp.type_store.put(
+                comp.gpa,
+                @unionInit(TypeStore.Type, @tagName(tag), bt.*),
+            );
+        },
+        else => {
+            if (wip.current.target == null and qt.is(comp, .array)) {
+                attr.used_as_type_attr = false;
+                return;
+            }
+
+            try wip.err(.invalid_nullability, .{qt});
+            return;
+        },
+    }
 }
 
 pub fn applyStmtAttrs(wip: *Wip, p: *Parser, stmt: Tree.Node.Index) !void {

--- a/src/aro/TypeStore.zig
+++ b/src/aro/TypeStore.zig
@@ -1793,6 +1793,7 @@ pub const Type = union(enum) {
 
     pub const Block = struct {
         func: QualType,
+        nullability: Pointer.Nullability = .default,
     };
 
     pub const Func = struct {

--- a/test/cases/ast/block types.c
+++ b/test/cases/ast/block types.c
@@ -326,3 +326,6 @@ function: 'fn (afunc: *fn (int) double, ablock: block (int) double) void'
 
     implicit return_stmt: 'void'
 
+typedef: 'block () int'
+ name: NullableBlock
+

--- a/test/cases/block types.c
+++ b/test/cases/block types.c
@@ -55,6 +55,9 @@ void my_func_accepting_block(double (*afunc)(int), double (^ablock)(int)) {
   __auto_type ^a;
 }
 
+// blocks can be nullable
+typedef int (^_Nullable NullableBlock)(void);
+
 /** manifest:
 syntax
 args = -fblocks -Wpedantic -Wno-gnu-auto-type --target=aarch64-macos

